File: sdk/ios/HotUpdateTemplateProvider.swift
import Foundation

/**
 * Template Provider with Hot Update support for iOS
 *
 * Replace your existing TemplateProvider with this one to enable hot updates.
 *
 * Usage:
 * ```swift
 * // Initialize hot update in AppDelegate
 * LynxHotUpdate.shared.initialize(deploymentKey: "your-key", serverUrl: "https://your-server.com")
 *
 * // Use HotUpdateTemplateProvider
 * let provider = HotUpdateTemplateProvider()
 * ```
 */
class HotUpdateTemplateProvider: NSObject, LynxTemplateProvider {
    
    func loadTemplate(withUrl url: String!, onComplete callback: LynxTemplateLoadBlock!) {
        DispatchQueue.global(qos: .userInitiated).async {
            // Try to load from hot update first
            if let data = LynxHotUpdate.shared.loadBundle(url) {
                DispatchQueue.main.async {
                    callback(data, nil)
                }
                return
            }
            
            // Fallback to bundle
            if let filePath = Bundle.main.path(forResource: url, ofType: "bundle") {
                do {
                    let data = try Data(contentsOf: URL(fileURLWithPath: filePath))
                    DispatchQueue.main.async {
                        callback(data, nil)
                    }
                } catch {
                    DispatchQueue.main.async {
                        callback(nil, error)
                    }
                }
            } else {
                let error = NSError(
                    domain: "com.lynx.hotupdate",
                    code: 404,
                    userInfo: [NSLocalizedDescriptionKey: "Bundle not found: \(url ?? "")"]
                )
                DispatchQueue.main.async {
                    callback(nil, error)
                }
            }
        }
    }
}

File: sdk/android/HotUpdateTemplateProvider.kt
package com.lynx.hotupdate

import android.content.Context
import com.lynx.tasm.provider.AbsTemplateProvider
import java.io.ByteArrayOutputStream

/**
 * Template Provider with Hot Update support
 * 
 * Replace your existing TemplateProvider with this one to enable hot updates.
 * 
 * Usage in your Application class:
 * ```kotlin
 * // Initialize hot update
 * LynxHotUpdate.init(this, "your-deployment-key", "https://your-server.com")
 * 
 * // Use HotUpdateTemplateProvider instead of your custom provider
 * LynxEnv.inst().setTemplateProvider(HotUpdateTemplateProvider(this))
 * ```
 */
class HotUpdateTemplateProvider(context: Context) : AbsTemplateProvider() {
    
    private val mContext: Context = context.applicationContext
    
    override fun loadTemplate(uri: String, callback: Callback) {
        Thread {
            try {
                val bundleBytes = LynxHotUpdate.loadBundle(uri)
                callback.onSuccess(bundleBytes)
            } catch (e: Exception) {
                // Fallback to assets if hot update fails
                try {
                    mContext.assets.open(uri).use { inputStream ->
                        ByteArrayOutputStream().use { outputStream ->
                            val buffer = ByteArray(1024)
                            var length: Int
                            while (inputStream.read(buffer).also { length = it } != -1) {
                                outputStream.write(buffer, 0, length)
                            }
                            callback.onSuccess(outputStream.toByteArray())
                        }
                    }
                } catch (fallbackError: Exception) {
                    callback.onFailed(fallbackError.message)
                }
            }
        }.start()
    }
}
